relay-server/src/actors/processor.rs

```diff
@@ -950,68 +992,251 @@ impl EnvelopeProcessorService {
+                dynamic_sampling::normalize(state);
+                event::filter(state)?;
+                dynamic_sampling::run(state, &self.inner.config);
 
-            if_processing!({
-                event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
-            });
-        }
+                // We avoid extracting metrics if we are not sampling the event while in non-processing
+                // relays, in order to synchronize rate limits on indexed and processed transactions.
+                if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
+                    self.extract_metrics(state)?;
+                }
 
-        if_processing!({
-            self.enforce_quotas(state)?;
-            profile::process(state, &self.inner.config);
-            self.process_check_ins(state);
-            span::process(state, self.inner.config.clone());
-        });
+                dynamic_sampling::sample_envelope(state)?;
 
-        if state.has_event() {
-            event::scrub(state)?;
-            event::serialize(state)?;
-            if_processing!({
-                span::extract_from_event(state);
-            });
-        }
+                if_processing!({
+                    event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    self.enforce_quotas(state)?;
+                    profile::process(state, &self.inner.config);
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                    if_processing!({
+                        span::extract_from_event(state);
+                    });
+                }
+
+                attachment::scrub(state);
+            }
+            ItemType::Profile => {
+                profile::filter(state);
+            }
+
+            // Standalone attachments.
+            ItemType::Attachment | ItemType::FormData => {
+                // Only some attachments types have create event set to true.
+                if state.creates_event() {
+                    event::extract(state, &self.inner.config)?;
+
+                    if_processing!({
+                        attachment::create_placeholders(state);
+                    });
+
+                    event::finalize(state, &self.inner.config)?;
+                    self.light_normalize_event(state)?;
+                    event::filter(state)?;
+
+                    if_processing!({
+                        event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    });
+                }
+
+                if_processing!({
+                    self.enforce_quotas(state)?;
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                }
+
+                attachment::scrub(state);
```

Comment:
nit: Should we put every one of these match arms into separate functions?

Reply:
It's a good question. 
I personally wouldn't do that, because it would require more changes, like refactoring `if_processing!` macro somehow to make sure it works outside of the `process_state` function. 
But I will have a look not, how can I do it better.